sdks/python/apache_beam/testing/benchmarks/nexmark/queries/query3.py

```diff
@@ -0,0 +1,164 @@
+
+We'll implement this query to allow 'new auction' events to come before the
+'new person' events for the auction seller. Those auctions will be stored until
+the matching person is seen. Then all subsequent auctions for a person will use
+the stored person record.
+"""
+
+from __future__ import absolute_import
+
+import logging
+
+import apache_beam as beam
+from apache_beam.coders import coders
+from apache_beam.testing.benchmarks.nexmark.models import name_city_state_id
+from apache_beam.testing.benchmarks.nexmark.models import nexmark_model
+from apache_beam.testing.benchmarks.nexmark.queries import nexmark_query_util
+from apache_beam.transforms import trigger
+from apache_beam.transforms import userstate
+from apache_beam.transforms import window
+from apache_beam.transforms.userstate import on_timer
+from apache_beam.utils.timestamp import Duration
+
+
+def load(events, metadata=None):
+  num_events_in_pane = 30
+  windowed_events = (
+      events
+      | beam.WindowInto(
+          window.GlobalWindows(),
+          trigger=trigger.Repeatedly(trigger.AfterCount(num_events_in_pane)),
+          accumulation_mode=trigger.AccumulationMode.DISCARDING,
+          allowed_lateness=Duration.of(0)))
+  auction_by_seller_id = (
+      windowed_events
+      | nexmark_query_util.JustAuctions()
+      | 'query3_filter_category' >> beam.Filter(lambda auc: auc.category == 10)
+      | 'query3_key_by_seller' >> beam.ParDo(
+          nexmark_query_util.AuctionBySellerFn()))
+  person_by_id = (
+      windowed_events
+      | nexmark_query_util.JustPerson()
+      | 'query3_filter_region' >> beam.Filter(
+          lambda person: person.state == 'OR' or person.state == 'ID' or person.
+          state == 'CA')
+      | 'query3_key_by_person_id' >> beam.ParDo(
+          nexmark_query_util.PersonByIdFn()))
+  return ({
+      nexmark_query_util.AUCTION_TAG: auction_by_seller_id,
+      nexmark_query_util.PERSON_TAG: person_by_id,
+  }
+          | beam.CoGroupByKey()
+          | 'query3_join' >> beam.ParDo(
+              JoinFn(metadata.get('max_auction_waiting_time')))
+          | 'query3_output' >> beam.Map(
+              lambda t: name_city_state_id.NameCiyStateId(
+                  t[1].name, t[1].city, t[1].state, t[0].id)))
+
+
+class JoinFn(beam.DoFn):
+  """
+  Join auctions and person by person id and emit their product one pair at
+  a time.
+
+  We know a person may submit any number of auctions. Thus new person event
+  must have the person record stored in persistent state in order to match
+  future auctions by that person.
+
+  However we know that each auction is associated with at most one person, so
+  only need to store auction records in persistent state until we have seen the
+  corresponding person record. And of course may have already seen that record.
+  """
+
+  AUCTIONS = 'auctions_state'
+  PERSON = 'person_state'
+  PERSON_EXPIRING = 'person_state_expiring'
+
+  auction_spec = userstate.ReadModifyWriteStateSpec(
```

Comment:
Should this be a BagUserState instead, generally it's better to use BagUserState for iterables.

Reply:
wouldn't I need to have a for loop to use the `add()` to add a list of stuff and then it might be slower in that realm

Reply:
not necessarily, ReadModifyWriteState uses BagUserState of size 1 underneath, if you use ReadModifyWriteState the whole list gets serialized/deserialized each time as a single item you make a change, BagUserState allows you to just add the new items.

Reply:
Cool, will do that